Xamarin.Forms.Platform.iOS/Renderers/MediaElementRenderer.cs

```diff
@@ -0,0 +1,455 @@
+﻿using AVFoundation;
+using AVKit;
+using CoreMedia;
+using Foundation;
+using System;
+using System.IO;
+using UIKit;
+using Xamarin.Forms.Internals;
+
+namespace Xamarin.Forms.Platform.iOS
+{
+	public sealed class MediaElementRenderer : UIView, IVisualElementRenderer, IEffectControlProvider
+	{
+		MediaElement MediaElement { get; set; }
+		IMediaElementController Controller => MediaElement as IMediaElementController;
+
+#pragma warning disable 0414
+		VisualElementTracker _tracker;
+#pragma warning restore 0414
+
+		AVPlayerViewController _avPlayerViewController = new AVPlayerViewController();
+
+		NSObject _playToEndObserver;
+		NSObject _statusObserver;
+		NSObject _rateObserver;
+
+		VisualElement IVisualElementRenderer.Element => MediaElement;
+
+		UIView IVisualElementRenderer.NativeView => this;
+
+		UIViewController IVisualElementRenderer.ViewController => _avPlayerViewController;
+		
+		bool _idleTimerDisabled = false;
+
+		public MediaElementRenderer()
+		{
+			_playToEndObserver = NSNotificationCenter.DefaultCenter.AddObserver(AVPlayerItem.DidPlayToEndTimeNotification, PlayedToEnd);
+			_avPlayerViewController.View.Frame = Bounds;
+			_avPlayerViewController.View.AutoresizingMask = UIViewAutoresizing.FlexibleWidth | UIViewAutoresizing.FlexibleHeight;
+			AddSubview(_avPlayerViewController.View);
+		}
+		
+		void SetKeepScreenOn(bool value)
+		{
+			if (value)
+			{
+				if (!UIApplication.SharedApplication.IdleTimerDisabled)
+				{
+					_idleTimerDisabled = true;
+					UIApplication.SharedApplication.IdleTimerDisabled = true;
+				}
+			}
+			else if (_idleTimerDisabled)
+			{
+				_idleTimerDisabled = false;
+				UIApplication.SharedApplication.IdleTimerDisabled = false;
+			}
+		}
+
+		void UpdateSource()
+		{
+			if (MediaElement.Source != null)
+			{
+				AVAsset asset = null;
+				
+				var uriSource = MediaElement.Source as UriMediaSource;
+				if (uriSource != null)
+				{
+					if (uriSource.Uri.Scheme == "ms-appx")
+					{
+						// used for a file embedded in the application package
+						asset = AVAsset.FromUrl(NSUrl.FromFilename(uriSource.Uri.LocalPath.Substring(1)));
+					}
+					else if (uriSource.Uri.Scheme == "ms-appdata")
+					{
+						string filePath = string.Empty;
+
+						if (uriSource.Uri.LocalPath.StartsWith("/local"))
+						{
+							filePath = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.MyDocuments), uriSource.Uri.LocalPath.Substring(7));
```

Comment:
Maybe this should be in the `ApplicationData` path rather? https://docs.microsoft.com/en-us/dotnet/api/system.environment.specialfolder?view=netframework-4.7.2

Reply:
On iOS this will be the local app Documents folder. The other option would be the Library folder which is meant for app data which is never visible to the user and not cloud synced. I'm guessing the latter would be the best option since it avoids accidentally cloud syncing large data files...